src/fsharp/symbols/Exprs.fs

```diff
@@ -530,11 +586,67 @@ module FSharpExprConvert =
                 let argR = ConvExpr cenv env arg
                 E.ILFieldSet(None, typR, fspec.Name, argR) 
 
+            | TOp.ILAsm([ ], _), _, [arg] -> 
+                ConvExprPrim cenv env arg
+
+            | TOp.ILAsm([ I_box _ ], _), [ty], [arg] -> 
+                let op = mkCallBox cenv.g m ty arg
+                ConvExprPrim cenv env op
+
+            | TOp.ILAsm([ I_unbox_any _ ], _), [ty], [arg] -> 
+                let op = mkCallUnbox cenv.g m ty arg
+                ConvExprPrim cenv env op
 
-            | TOp.ILAsm([ AI_ceq ], _), _, [arg1;arg2]  -> 
+            | TOp.ILAsm([ I_isinst _ ], _), [ty], [arg] -> 
+                let op = mkCallTypeTest cenv.g m ty arg
+                ConvExprPrim cenv env op
+
+            | TOp.ILAsm([ I_ldtoken _ ], _), [ty], _ -> 
+                let op = mkCallTypeOf cenv.g m ty
```

Comment:
``ldtoken`` only gets the runtime handle for the type.  To invert properly to``typeof`` you need to consider the whole expression produced by uses of this:

        let inline typeof<'T> =
            let tok = (# "ldtoken !0" type('T) : System.RuntimeTypeHandle #)
            System.Type.GetTypeFromHandle(tok)

Reply:
There is one [here](https://github.com/forki/visualfsharp/blob/f352696f7a0c6956c430ec58bf9a9db39a02d076/tests/service/ExprTests.fs#L286) with this result:
- Unoptimized:
`let tyfuncEx1 = Operators.TypeOf<'T> ()`
- Optimized:
`let tyfuncEx1 = Type.GetTypeFromHandle (Operators.TypeOf<'T> ())`

Reply:
@ncave Right, that optimized code looks incorrectly typed, since TypeOf returns a System.Type, and GetTypeFromHandle takes a RuntimeTypeHandle

Reply:
@dsyme Is it ok to do just TypeOf then?

Reply:
@dsyme latest removes GetTypeFromHandle.